File: go/leetcode/238. product of array except self.go
func productExceptSelf(nums []int) []int {
    ans := make([]int, len(nums))
    forward := make([]int, len(nums))
    backward := make([]int, len(nums))
    
    // initialize 
    forward[0] = 1
    backward[len(nums) - 1] = 1
    
    // move forward
    for i := 0; i < len(nums) - 1; i++ {
        forward[i + 1] = nums[i] * forward[i]
    }

    
    // move backward
    for i := len(nums) - 1; i > 0; i-- {
        backward[i - 1] = nums[i] * backward[i]
    }    
    

    for i := 0; i < len(nums); i++ {
        ans[i] = forward[i] * backward[i]
    }
    
    return ans
    
}
  

File: c++/leetcode/238. Product of Array Except Self.cpp
class Solution
{
public:
  vector<int> productExceptSelf(vector<int> &nums)
  {
    vector<int> ans(nums.size());
    const int sz = nums.size();

    ans[0] = 1;

    for (int i = 1; i < sz; ++i)
    {
      ans[i] = nums[i - 1] * ans[i - 1];
    }

    int prod = 1;
    for (int i = sz - 1; i >= 0; --i)
    {
      ans[i] = prod * ans[i];
      prod *= nums[i];
    }

    return ans;
  }
};